src/core/lib/iomgr/tcp_posix.c

```diff
@@ -423,11 +423,18 @@ static bool tcp_flush(grpc_tcp *tcp, grpc_error **error) {
       }
     }
 
+    size_t bytes_written = 0;
+    for (iov_size = unwind_slice_idx; iov_size < tcp->outgoing_slice_idx; iov_size++) {
+        bytes_written += GRPC_SLICE_LENGTH(
+          tcp->outgoing_buffer->slices[iov_size]);
+        if(bytes_written <= (size_t) sent_length) {
+          grpc_slice_write(exec_ctx, tcp->outgoing_buffer->slices[iov_size]);
```

Comment:
Good proof of concept it seems...

Endpoints take ownership of the write slice buffer during the write, and make no guarantees about its contents after the write...

So we should be able to get this optimization without a special slice type just by being better about mutating that slice buffer in this code (unref as we go, clear at completion)